### PR TITLE
Containers lib image handling

### LINKED ISSUE #35: Use containers/image library for docker-less pull and mount
Requiring access to the docker.socket is a lot of privilege. With the https://github.com/containers/image/ library this should not be required.

### PR DESCRIPTION
**Pull request by enoodle**
This will use github.com/containers/image to pull images instead of connecting to the docker socket (which requires running in privileged mode).

Use of the docker daemon is still possible via --use-docker

Fixes #35

(this PR is an update of #58 rebased on master. Lets make this the last time.. )

### CODE REVIEW

**Comment by enoodle** [comment]
@legionus @mfojtik @zgalor Can you review ?

File: Dockerfile [hunk 1]
```diff
@@ -2,15 +2,23 @@ FROM centos:7
...
     yum clean all
 
+ENV PATH=$PATH:/usr/local/go/bin
+ENV CGO_ENABLED=0
+
 COPY .  /go/src/github.com/openshift/image-inspector
 
-RUN GOBIN=/usr/bin \
+RUN cd /tmp && \
+    curl -LO https://storage.googleapis.com/golang/go1.7.linux-amd64.tar.gz && \
```
Comment (zgalor) [inline Dockerfile]: I think you should install via yum. centos repo already has golang-1.8.3-1.el7.x86_64 available

File: kubernetes/image-inspector-app.yaml [hunk 2]
```diff
@@ -1,26 +1,54 @@
...
+  users:
+  - system:serviceaccount:${PROJECT}:${SERVICEACCOUNT}
+  defaultAddCapabilities:
+    - SYS_CHROOT
+- apiVersion: v1
+  kind: Pod
+  metadata:
+    name: image-inspector
+    labels:
+      name: inspect-fedora-latest
```
Comment (zgalor) [inline kubernetes/image-inspector-app.yaml]: label `inspect-fedora-latest` is inconsistent with pod name `image-inspector`

File: pkg/inspector/image-inspector.go [hunk 3]
```diff
@@ -126,45 +128,61 @@ func (i *defaultImageInspector) Inspect() error {
...
 
-		randomName, err := generateRandomName()
-		if err != nil {
-			return err
+			if err := i.extractDownloadedImage(inspectInfo.Layers); err != nil {
+				return fmt.Errorf("extracting downloaded image: %v ", err)
+			}
+
+			i.meta.Image = inspectInfoToDockerImage(inspectInfo, imageDigest)
 		}
```
Comment (zgalor) [inline pkg/inspector/image-inspector.go]: i think this line is missing here:
`scanResults.ImageID = i.meta.Image.ID		`

File: pkg/inspector/image-inspector.go [hunk 4]
```diff
@@ -244,7 +263,7 @@ func (i *defaultImageInspector) Inspect() error {
 
 	if len(i.opts.PostResultURL) > 0 {
 		if err := i.postResults(scanResults); err != nil {
-			log.Printf("Error posting results: %v", err)
+			log.Printf("posting results: %v", err)
```
Comment (zgalor) [inline pkg/inspector/image-inspector.go]: looks like `Error` prefix should stay here

File: pkg/inspector/image-inspector.go [hunk 5]
```diff
@@ -636,16 +590,188 @@ func createOutputDir(dirName string, tempName string) (string, error) {
...
 	}
 	return dirName, nil
 }
+
+// pullImage pulls the image using containers/image library
+// to directly connect to the docker registry
+// and will return its metadata (*types.ImageInspectInfo).
+// Assumes the image to be pulled is with a transport prefix
+// or if not attempts to add the "docker://" prefix to the image.
+func (i *defaultImageInspector) pullImage() (*types.ImageInspectInfo, digest.Digest, error) {
```
Comment (zgalor) [inline pkg/inspector/image-inspector.go]: I suggest changing to `containersPullImage` and change `dockerPullImage` to 'dockerSocketPullImage`

File: pkg/inspector/image-inspector.go [hunk 6]
```diff
@@ -636,16 +590,188 @@ func createOutputDir(dirName string, tempName string) (string, error) {
...
+		destDir := i.opts.DstPath
+		log.Printf("Untar %s\n", filename)
+		if err := util.UntarGzLayer(filename, destDir); err != nil {
+			return fmt.Errorf("extracting gzipped layer tarball: %v", err)
+		}
+	}
+	return nil
+}
+
+// dockerCertPath will try to extract the registry name from the image and return
```
Comment (zgalor) [inline pkg/inspector/image-inspector.go]: s/dockerCertPath/certPath/g

File: pkg/inspector/image-inspector.go [hunk 7]
```diff
@@ -636,16 +590,188 @@ func createOutputDir(dirName string, tempName string) (string, error) {
...
+	policy := &signature.Policy{
+		Default: []signature.PolicyRequirement{signature.NewPRInsecureAcceptAnything()},
+	}
+	policyContext, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return nil, "", fmt.Errorf("creating context for policy %v: %v", policy, err)
+	}
+	defer policyContext.Destroy()
+
+	srcRef, err := alltransports.ParseImageName(i.opts.Image)
```
Comment (zgalor) [inline pkg/inspector/image-inspector.go]: in this block I would check that `i.opts.Image` fits the `registry://name` pattern, and if the registry is missing add `docker://` before sending to ParseImageName. 

This pattern is also required for `certPath` function later

File: pkg/util/util.go [hunk 8]
```diff
@@ -22,3 +41,105 @@ func StringInList(s string, l []string) bool {
...
+
+func ExtractLayerTar(src io.Reader, destination string) error {
+	tr := tar.NewReader(src)
+	for {
+		hdr, err := tr.Next()
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return fmt.Errorf("Unable to read tar: %v\n", err)
```
Comment (zgalor) [inline pkg/util/util.go]: better be consistent with previous changes of error formatting: `reading tar`

File: pkg/util/util.go [hunk 9]
```diff
@@ -22,3 +41,105 @@ func StringInList(s string, l []string) bool {
...
+		if strings.HasPrefix(filepath.Base(dstpath), whiteoutFilePrefix) {
+			os.RemoveAll(dstpath)
+			continue
+		}
+
+		switch hdr.Typeflag {
+		case tar.TypeDir:
+			if err := os.Mkdir(dstpath, mode); err != nil {
+				if !os.IsExist(err) {
+					return fmt.Errorf("Unable to create directory: %v", err)
```
Comment (zgalor) [inline pkg/util/util.go]: see prev

File: pkg/util/util.go [hunk 10]
```diff
@@ -22,3 +41,105 @@ func StringInList(s string, l []string) bool {
...
+				}
+				err = os.Chmod(dstpath, mode)
+				if err != nil {
+					return fmt.Errorf("Unable to update directory mode: %v", err)
```
Comment (zgalor) [inline pkg/util/util.go]: here also

File: pkg/inspector/image-inspector.go (hunk 5 above)
Comment (enoodle) [reply pkg/inspector/image-inspector.go]: :+1:  for containersPullImage but I think I will leave `dockerPullImage` as is.

**Comment by simon3z** [comment]
@enoodle can you rebase?
@rhdedgar can you review this?

**Comment by enoodle** [comment]
@simon3z I rebased this but I think it will be better to go with https://github.com/openshift/image-inspector/pull/95 (and preferably after rebasing https://github.com/openshift/image-inspector/pull/90)

**Comment by simon3z** [comment]
> @simon3z I rebased this but I think it will be better to go with #95 (and preferably after rebasing #90)

OK for me (rebase those). Please close the PRs that are not relevant then, I think both @oourfali and I got the impression that this was the one to review.

**Comment by enoodle** [comment]
Closing in favor of https://github.com/openshift/image-inspector/pull/95

**Pull request closed without merging**